pylops/waveeqprocessing/_kirchhoff_cuda.py

```diff
@@ -0,0 +1,605 @@
+                num_sources_for_stream = sources_per_stream + (
+                    1 if i < remainder else 0
+                )
+                sources_for_stream = list(
+                    range(source_counter, source_counter + num_sources_for_stream)
+                )
+                self.sources_per_streams[stream] = sources_for_stream
+                source_counter += num_sources_for_stream
+            n_runs = (
+                num_streams * self.ns * self.ni * self.nr
+            )  # number_of_times_to_run_kernel
+            self.num_threads_per_blocks = current_device.WARP_SIZE * 8
+            self.num_blocks = (
+                n_runs + (self.num_threads_per_blocks - 1)
+            ) // self.num_threads_per_blocks
+        else:
+            warp = current_device.WARP_SIZE
+            self.num_threads_per_blocks = (warp, warp)
+            self.num_blocks = (
+                (self.ns + self.num_threads_per_blocks[0] - 1)
+                // self.num_threads_per_blocks[0],
+                (self.nr + self.num_threads_per_blocks[1] - 1)
+                // self.num_threads_per_blocks[1],
+            )
+
+    def _data_prep_dynamic(
+        self,
+        ns,
+        nr,
+        nt,
+        ni,
+        nz,
+        dt,
+        aperture,
+        angleaperture,
+        aperturetap,
+        vel,
+        six,
+        rix,
+        trav_recs,
+        angle_recs,
+        trav_srcs,
+        angle_srcs,
+        amp_srcs,
+        amp_recs,
+    ):
+        """Data preparation for dynamic computation
+
+        Prepare data for dynamic computation by transfering some variables to device
+        memory in advance once.
+        """
+        ns_d = np.int32(ns)
+        nr_d = np.int32(nr)
+        nt_d = np.int32(nt)
+        ni_d = np.int32(ni)
+        nz_d = np.int32(nz)
+        dt_d = np.float32(dt)
+        aperturemin_d = np.float32(aperture[0])
+        aperturemax_d = np.float32(aperture[1])
+        angleaperturemin_d = np.float32(angleaperture[0])
+        angleaperturemax_d = np.float32(angleaperture[1])
+        aperturetap_d = cuda.to_device(aperturetap)
+        vel_d = cuda.to_device(vel)
+        six_d = cuda.to_device(six)
+        rix_d = cuda.to_device(rix)
+        self.const_inputs = (
+            ns_d,
+            nr_d,
+            nt_d,
+            ni_d,
+            nz_d,
+            dt_d,
+            aperturemin_d,
+            aperturemax_d,
+            angleaperturemin_d,
+            angleaperturemax_d,
+            vel_d,
+            aperturetap_d,
+            six_d,
+            rix_d,
+        )
+
+        self.trav_recs_d_global = cuda.to_device(trav_recs)
+        self.angles_recs_d_global = cuda.to_device(angle_recs)
+        self.trav_srcs_d_global = cuda.to_device(trav_srcs)
+        self.angles_srcs_d_global = cuda.to_device(angle_srcs)
+        self.amp_srcs_d_global = cuda.to_device(amp_srcs)
+        self.amp_recs_d_global = cuda.to_device(amp_recs)
```

Comment:
I'm not sure how much overhead these `to_device` are adding, but if you'd like to make this more efficient, you can queue the memory transfers per stream and use pinned arrays to ensure async. Here's a snippet (untested):

```python
streams = list(self.sources_per_streams.keys())
n = len(streams)
with cuda.pinned(
    aperturetap,
    vel,
    six,
    rix,
    trav_recs,
    angle_recs,
    trav_srcs,
    angle_srcs,
    amp_srcs,
    amp_recs,
):
    aperturetap_d = cuda.to_device(aperturetap, stream=streams[0 % n])
    vel_d = cuda.to_device(vel, stream=streams[1 % n])
    six_d = cuda.to_device(six, stream=streams[2 % n])
    rix_d = cuda.to_device(rix, stream=streams[3 % n])
    self.trav_recs_d_global = cuda.to_device(trav_recs, stream=streams[4 % n])
    self.angles_recs_d_global = cuda.to_device(
        angle_recs, stream=streams[5 % n]
    )
    self.trav_srcs_d_global = cuda.to_device(trav_srcs, stream=streams[6 % n])
    self.angles_srcs_d_global = cuda.to_device(
        angle_srcs, stream=streams[7 % n]
    )
    self.amp_srcs_d_global = cuda.to_device(amp_srcs, stream=streams[8 % n])
    self.amp_recs_d_global = cuda.to_device(amp_recs, stream=streams[9 % n])
cuda.synchronize()
```